core/src/main/java/com/graphhopper/routing/weighting/AbstractWeighting.java

```diff
@@ -36,18 +36,20 @@
     protected final DecimalEncodedValue avSpeedEnc;
     protected final BooleanEncodedValue accessEnc;
     private final TurnCostProvider turnCostProvider;
+    private final String name;
 
-    protected AbstractWeighting(FlagEncoder encoder) {
-        this(encoder, NO_TURN_COST_PROVIDER);
+    protected AbstractWeighting(String name, FlagEncoder encoder) {
```

Comment:
We can now pass a name to AbstractWeighting, while previously this was hard-coded in the subclasses. As long as the subclasses hard-code this name in their constructor instead of overriding `getName()` this is more or less equivalent. But its also a bit ugly because 'intermediate' subclasses like FastestWeighting expose this parameter in their constructor and why would I call the FastestWeighting something different than "fastest"? I understand that the custom weighting needs a name here (but maybe only as a workaround as long as there is no profile parameter?). Maybe it would be sufficient to just have the name parameter in CustomWeighting's constructor?

Reply:
Maybe there is a different solution with the profile name PR, but for now this was the only possibility IMO.

Reply:
But why is the name field not in CustomWeighting?

Reply:
What do you mean here? We pass `name=key(name)` to the super class of CustomWeighting.

Reply:
https://github.com/graphhopper/graphhopper/pull/1841/files#diff-597b60bb826bc11b659858a59d5112d3R72

Reply:
Yes I and I am wondering why. If we did not pass it to AbstractWeighting we could just keep it in CustomWeighting (and there would not be a need to have `String name` in the `FastestWeighting` constructor for example. To me it looks like the constructor of `FastestWeighting` only takes a name parameter because of `PriorityWeighting`. If I was using `FastestWeighting` I would be confused why it takes a name via its constructor. But well this is maybe just me.

Reply:
Ah, ok. Will try to revert this new constructor parameter.

Reply:
Ah, the problem with this is that we have a verification `!isValidName(getName())` in AbstractWeighting. And because the call to "super"-constructor must be the first call the `getName()` returns null. Should we move the isValidName-check out of the AbstractWeighting class or keep this new `name` parameter?

Reply:
Ah ok, having a 'valid' name is important if the weighting name is used as url-parameter or file name. On the other hand its hard-coded for most weightings and even for `CustomWeighting` the 'name' of the weighting should soon be irrelevant, because only the profile name will be important. I think I would rather remove it.

Reply:
With the latest master (including your profiles work) I was able to remove this from AbstractWeighting and this also cleaned up a few other places (nice!). 

Still I had to introduce a GHRequest.getProfile helper method. See the latest commit after the merge. Is this a good idea?

Reply:
Nice you got rid of the name and this cleaned things up. Yes of course GHRequest needs get/setProfile. Actually I am working on this right now.

Reply:
Btw so far I did not consider that we are using `vehicle=pt` for PT on the `/route` endpoint? How is this going to fit into our new way of selecting the profile? Should we (behind the converter) change this to `profile=pt`?

Reply:
For example this test: https://github.com/graphhopper/graphhopper/blob/eb6e1d104ca547326120682c1299415045a5d616/web/src/test/java/com/graphhopper/http/resources/GtfsTest.java#L75-L86

Reply:
Interesting. When I think more about it then public transit (or multi modal routing) has more similarity to CustomWeighting than to a pure FlagEncoder / vehicle. E.g. it should be able to specify profiles it can use and also we have to ensure that it cannot be used for CH or LM 

```yml
profile:
   name: pt
   vehicle: pt # we could use this special vehicle to avoid the LM/CH preparations
   additional_profiles: foot, bike # or wheelchair or even car for free-floating carsharing.
```